Tune maximum thread count for streaming dataflow worker executor dynamically.

MelodyShen opened the pull request:
Workers will read the StreamignScalingReportResponse from worker messages and configure the executor pool size based on the specified value.

R: @scwhittle 

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://github.com/apache/beam/blob/master/CONTRIBUTING.md#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[![Build python source distribution and wheels](https://github.com/apache/beam/workflows/Build%20python%20source%20distribution%20and%20wheels/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[![Python tests](https://github.com/apache/beam/workflows/Python%20tests/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[![Java tests](https://github.com/apache/beam/workflows/Java%20Tests/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[![Go tests](https://github.com/apache/beam/workflows/Go%20tests/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI or the [workflows README](https://github.com/apache/beam/blob/master/.github/workflows/README.md) to see a list of phrases to trigger workflows.

scwhittle commented on runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/util/BoundedQueueExecutor.java:
```diff
@@ -48,6 +49,7 @@ public BoundedQueueExecutor(
       int maximumElementsOutstanding,
       long maximumBytesOutstanding,
       ThreadFactory threadFactory) {
+    this.maximumThreadCount.set(maximumPoolSize);
```
how about 
this.maximumThreadCount = new AtomicInteger(maximumPoolSize);
instead of assignment above

MelodyShen replied:
Changed to int and use synchronized instead.

scwhittle commented on runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/util/BoundedQueueExecutor.java:
```diff
@@ -70,7 +72,7 @@ protected void beforeExecute(Thread t, Runnable r) {
             super.afterExecute(r, t);
             synchronized (this) {
-              if (activeCount.getAndDecrement() == maximumPoolSize) {
+              if (activeCount.getAndDecrement() == maximumThreadCount.get()) {
```
I think this might need to be
if (activeCount.getAndDecrement() <= maximumThreadCount.get() &&
    startTimeMaxActiveThreadsUsed > 0)

you could maybe test this by adding more work to queue than the limit, having the work that is scheduled block, then increase the limit and let all the work complete.  We should see totalTimeMaxActiveThreadsUsed incremented

Another way to fix it could be to modify the setMaximumPoolSize to reset and count startTimeMaxActiveThreadsUsed if the limit increased (it will be restarted shortly as new threads start and call beforeExecute).

MelodyShen replied:
Yeah that makes sense.

scwhittle commented on runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/util/BoundedQueueExecutor.java:
```diff
@@ -104,6 +106,20 @@ public void forceExecute(Runnable work, long workBytes) {
   }
 
+  // Set the maximum/core pool size of the executor.
+  public void setMaximumPoolSize(int maximumPoolSize) {
```
sychronized?
seems safer when comparing the atomic with other things like pool size

Perhaps the atomics in this class should just be removed and protected by synchronized.  Using the atomics within synchronized block is more overhead and these accessors reading the atomics are not called very frequently I believe.

MelodyShen replied:
Sounds good. I will refactor the class for related atomics and non-guarded fields.

scwhittle commented on runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/util/BoundedQueueExecutor.java:
```diff
@@ -123,6 +139,10 @@ public int activeCount() {
   }
 
+  public int maximumThreadCount() {
+    return maximumThreadCount.intValue();
```
github won't let me add a comment, but above allThreadsActiveTime shoudl be synchronized so it isn't racy reading that while other threads are modifying it.

can add a @GuardedBy annotation to totalTimeMaxActiveThreadsUsed

scwhittle commented on runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/util/BoundedQueueExecutor.java:
```diff
@@ -38,6 +38,7 @@ public class BoundedQueueExecutor {
   private int elementsOutstanding = 0;
   private long bytesOutstanding = 0;
   private final AtomicInteger activeCount = new AtomicInteger();
+  private final AtomicInteger maximumThreadCount = new AtomicInteger();
```
would be good to add a unit test covering the new methods to this class directly

MelodyShen replied:
Sure.

scwhittle commented on runners/google-cloud-dataflow-java/worker/src/test/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorkerTest.java:
```diff
@@ -2916,6 +2916,69 @@ public void testActiveThreadMetric() throws Exception {
+  @Test
+  public void testOverrideMaximumThreadCount() throws Exception {
+    int maxThreads = 5;
+    int threadExpirationSec = 60;
```
this doesn't matter right? If so set it higher so it isn't possible that it triggers.

MelodyShen replied:
Cool!

scwhittle commented on runners/google-cloud-dataflow-java/worker/src/test/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorkerTest.java:
```diff
@@ -2916,6 +2916,69 @@ public void testActiveThreadMetric() throws Exception {
+    AtomicBoolean stop = new AtomicBoolean(false);
+    // setting up actual implementation of executor instead of mocking to keep track of
+    // active thread count.
+    BoundedQueueExecutor executor =
```
the test that BoundedQueueExecutor works when calling setMaximumPoolSize seems like it should be in BoundedQueueExecutorTest.

The test at the streaming dataflow worker level seems like it should have the work update client and validate plumbing of responses to setting the max on the queue.  The queue itself could be a mock in that test.

MelodyShen replied:
Sounds good. Will refactor it.

MelodyShen reviewed (commented):
Thanks for feedback! Working on addressing them.

MelodyShen commented:
Hi @scwhittle sorry I rebased the branch incorrectly which caused the previous commit history got lost. The main changes since last review are:
1. BoundedQueueExecutor: Use synchronized instead of Atomic for all the thread-unsafe fields.
2. StreamingDataflowWorkerTest: update test
3. Added a new test file BoundedQueueExecutorTest to test behaviors related to maximumThreadCount and TotalMaxThreadActiveTimeUsed.

Thanks for your time!

scwhittle commented on runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java:
```diff
@@ -500,8 +503,10 @@ static StreamingDataflowWorker forTesting(
-      Function<String, ScheduledExecutorService> executorSupplier) {
-    BoundedQueueExecutor boundedQueueExecutor = createWorkUnitExecutor(options);
+      Function<String, ScheduledExecutorService> executorSupplier,
+      BoundedQueueExecutor executor) {
```
how about naming workUnitExecutor and can just use same variable below to avoid proliferation of names

if (workUnitExeuctor == null) workUnitExecutor = createWorkUnitExecutor(options);

scwhittle commented on runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java:
```diff
@@ -1918,6 +1926,26 @@ private Optional<WorkerMessage> createWorkerMessageForPerWorkerMetrics() {
+  private void readAndSaveWorkerMessageResponseForStreamingScalingReportResponse(
+      List<WorkerMessageResponse> responses) {
+    for (WorkerMessageResponse response : responses) {
+      if (response.getStreamingScalingReportResponse() != null) {
```
would it be possible to get multiple responses? If so perhaps it would be best to find the last value in the loop and then just set once instead of possibly setting multiple times.

scwhittle commented on runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/util/BoundedQueueExecutor.java:
```diff
@@ -37,8 +37,17 @@ public class BoundedQueueExecutor {
   private final Monitor monitor = new Monitor();
   private int elementsOutstanding = 0;
```
comment that elementsOutstanding and bytesOutstanding are guarded by the monitor

scwhittle commented on runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/util/BoundedQueueExecutor.java:
```diff
@@ -59,8 +69,8 @@ public BoundedQueueExecutor(
-              if (activeCount.getAndIncrement() >= maximumPoolSize - 1) {
+            synchronized (BoundedQueueExecutor.this) {
+              if (activeCount++ >= maximumThreadCount - 1) {
                 startTimeMaxActiveThreadsUsed = System.currentTimeMillis();
```
only reset if zero? I think it could be racy since

scwhittle commented on runners/google-cloud-dataflow-java/src/test/java/org/apache/beam/runners/dataflow/util/BoundedQueueExecutorTest.java:
```diff
@@ -0,0 +1,189 @@
+/*
```
I think this file should be in the directory path matching other util files like:

runners/google-cloud-dataflow-java/worker/src/test/java/org/apache/beam/runners/dataflow/worker/util/MemoryMonitorTest.java

scwhittle reviewed (changes requested):
No worries about losing previous on rebase, github is annoying that way. I haven't figured out a good way to do that other than not rebasing/merging until the end which isn't great.

scwhittle commented on runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java:
```diff
@@ -653,6 +659,9 @@ private static void setUpWorkLoggingContext(String workId, String computationId)
   }
 
   private int chooseMaximumNumberOfThreads() {
+    if (maxThreadCountOverride.get() != 0) {
```
since this is possibly modified, it's safer to save the value you observe instead of observing it again

probably wouldn't ever actually change between calls but might as well be safe.

scwhittle commented on runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/util/BoundedQueueExecutor.java:
```diff
@@ -115,12 +143,16 @@ public boolean executorQueueIsEmpty() {
+    return maximumThreadCount;
   }
 
   public long bytesOutstanding() {
```
doesn't look modified by this PR but bytesOutstanding() and elementsOutstanding() don't seem safe to read here if they are not beneath the monitor

can these public methods be removed? or shoudl they be changed to use monitor?

MelodyShen replied:
Yeah they are not thread safe. They are currently used in StreamingDataflowWorker to gather metrics so it seems okay to me to leave them unsafe but in case we use them for other purposes, I feel it is better to be changed to use monitor.

scwhittle commented on runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/util/BoundedQueueExecutor.java:
```diff
@@ -149,18 +191,22 @@ public String summaryHtml() {
 
       builder.append("Work Queue Size: ");
-      builder.append(elementsOutstanding);
+      builder.append(elementsOutstanding());
```
should the monitor.enter() at the top of summaryHtml be removed if using the accessors that grab the monitor?

not sure if it is reentrant. Would be good to add renderHtml to the unit test either way.

MelodyShen replied:
Ah I didn't notice that. Reverted and added a test case for it.

scwhittle commented on runners/google-cloud-dataflow-java/worker/src/test/java/org/apache/beam/runners/dataflow/worker/util/BoundedQueueExecutorTest.java:
```diff
@@ -0,0 +1,196 @@
+          latch.countDown();
+          int count = 0;
+          while (!stop.get()) {
+            count += 1;
```
add Thread.sleep(10) or something to avoid burning cpu

or could make stop a countdown latch as well to signal.

MelodyShen replied:
Sure change to use countdownlatch.

scwhittle commented on runners/google-cloud-dataflow-java/worker/src/test/java/org/apache/beam/runners/dataflow/worker/util/BoundedQueueExecutorTest.java:
```diff
@@ -0,0 +1,196 @@
+                .setDaemon(true)
+                .build());
+  }
+
```
add a basic test since we have a test for this now
below we never schedule new work after previous work started for example

MelodyShen replied:
Done.

scwhittle commented on runners/google-cloud-dataflow-java/worker/src/test/java/org/apache/beam/runners/dataflow/worker/util/BoundedQueueExecutorTest.java:
```diff
@@ -0,0 +1,232 @@
+    Runnable m3 = createSleepProcessWorkFn(processStart3, processStop3);
+
+    executor.execute(m1, 1);
+    assertTrue(processStart1.await(1000, TimeUnit.MILLISECONDS));
```
for all of these awaits where we expect it to happen, await without a timeout to prevent possible test flakiness. test deadline can be sufficient

MelodyShen replied:
Cool!

scwhittle commented on runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/util/BoundedQueueExecutor.java:
```diff
@@ -104,6 +117,21 @@ public void forceExecute(Runnable work, long workBytes) {
     executeLockHeld(work, workBytes);
```
can you rename executeLockHeld to executeMonitorHeld to clarify since we have both monitor and synchronized lock now?

scwhittle commented on runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/util/BoundedQueueExecutor.java:
```diff
@@ -59,8 +72,8 @@ public BoundedQueueExecutor(
-            synchronized (this) {
-              if (activeCount.getAndIncrement() >= maximumPoolSize - 1) {
+            synchronized (BoundedQueueExecutor.this) {
+              if (activeCount++ >= maximumThreadCount - 1 && startTimeMaxActiveThreadsUsed == 0) {
```
nit: how about
++activeCount >= maximumThreadCount

seems easier to read

MelodyShen replied:
Sure!

scwhittle commented on runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/util/BoundedQueueExecutor.java:
```diff
@@ -69,8 +82,8 @@ protected void beforeExecute(Thread t, Runnable r) {
-            synchronized (this) {
-              if (activeCount.getAndDecrement() == maximumPoolSize) {
+            synchronized (BoundedQueueExecutor.this) {
+              if (activeCount-- <= maximumThreadCount && startTimeMaxActiveThreadsUsed > 0) {
```
ditto,
--activeCount < maximumThreadCount 
seems simpler

scwhittle commented on runners/google-cloud-dataflow-java/worker/src/test/java/org/apache/beam/runners/dataflow/worker/util/BoundedQueueExecutorTest.java:
```diff
@@ -0,0 +1,232 @@
+    executor.execute(m3, 1);
+    assertFalse(processStart3.await(1000, TimeUnit.MILLISECONDS));
+
+    // Increase the max thread count
```
could you also verify that increasing the bytes would let something in similarly?

MelodyShen replied:
Currently the maximumBytesOutstanding is final and we don't change it at runtime. I added a test case for scheduling work when maximumBytesOutstanding is reached.

scwhittle commented on runners/google-cloud-dataflow-java/worker/src/test/java/org/apache/beam/runners/dataflow/worker/util/BoundedQueueExecutorTest.java:
```diff
@@ -0,0 +1,232 @@
+    assertEquals(0l, executor.allThreadsActiveTime());
+    stop.countDown();
+    while (executor.activeCount() != 0) {
+      // Waiting for all threads to be ended.
```
sleep a little here while polling to not burn cpu and slow down completion.

scwhittle commented on runners/google-cloud-dataflow-java/worker/src/test/java/org/apache/beam/runners/dataflow/worker/util/BoundedQueueExecutorTest.java:
```diff
@@ -0,0 +1,232 @@
+    executor.setMaximumPoolSize(5, 105);
+    stop.countDown();
+    while (executor.activeCount() != 0) {
+      // Waiting for all threads to be ended.
```
ditto

scwhittle commented on runners/google-cloud-dataflow-java/worker/src/test/java/org/apache/beam/runners/dataflow/worker/util/BoundedQueueExecutorTest.java:
```diff
@@ -0,0 +1,232 @@
+    executor.execute(m3, 1);
+    assertFalse(processStart3.await(1000, TimeUnit.MILLISECONDS));
+
+    assertEquals(0l, executor.allThreadsActiveTime());
```
this made me realize we're only updating the count after the active work completes.  That could delay possibly requesting additional threads if there is long-running work on the available threads.  Maybe in a followup we might want to periodically increase this with some monitoring thread instead of relying on the finishing threads themselves to increment.

MelodyShen replied:
Yeah if all the threads are occupied by the long-running tasks, we may under-estimate the related time counters. Not sure if we use these counters for any decisions now but if later we decide to tune the thread count based on these time counters and thread utilization, it might cause problem. Let me create a bug to follow up on this.

scwhittle reviewed (changes requested):
Sorry for the delay!

MelodyShen commented:
:runners:google-cloud-dataflow-java:examples-streaming:windmillPreCommit seems to be broken...

MelodyShen commented:
Rebased the master branch. The base code was refactored a lot recently so moved the related changes from StreamingDataflowWorker to StreamingWorkerStatusReporter based on it.

MelodyShen commented:
Hi @scwhittle thanks for reviewing the changes. I have rebased master to catch up with the latest and all checks passed. Would you mind merging the PR when available? Thanks!

scwhittle commented:
mac os failures appear unrelated
